ulib/LowStar.Regional.fst

```diff
@@ -0,0 +1,79 @@
+module LowStar.Regional
+
+open FStar.All
+open FStar.Integers
+open FStar.Classical
+open LowStar.Modifies
+
+module HH = FStar.Monotonic.HyperHeap
+module HS = FStar.HyperStack
+module HST = FStar.HyperStack.ST
+module MHS = FStar.Monotonic.HyperStack
+
+/// Regionality
+
+type erid = rid:HH.rid{HST.is_eternal_region rid}
+
+// Motivation: we want to ensure that all stateful operations for a value of
+// type `a` are within the `region_of` the value.
+noeq type regional a =
+| Rgl:
+    // The target type should have a region where it belongs.
+    region_of: (a -> GTot HH.rid) ->
+
+    // A stateless value of type `a`.
+    // It does not have to satisfy the invariant `r_inv` described below.
+    dummy: a ->
+
+    // A representation type of `a` and a corresponding conversion function
+    repr: Type0 ->
+    r_repr: (HS.mem -> a -> GTot repr) ->
+
+    // An invariant we want to maintain for each operation.
+    // For example, it may include `live` and `freeable` properties
+    // for related objects.
+    r_inv: (HS.mem -> a -> GTot Type0) ->
+    r_inv_reg:
+      (h:HS.mem -> v:a ->
+      Lemma (requires (r_inv h v))
+	    (ensures (MHS.live_region h (region_of v)))) ->
+
+    // A core separation lemma, saying that the invariant and represenation
+    // are preserved when an orthogonal state transition happens.
+    r_sep:
+      (v:a -> p:loc -> h:HS.mem -> h':HS.mem ->
+      Lemma (requires (r_inv h v /\
+		      loc_disjoint (loc_all_regions_from false (region_of v)) p /\
+		      modifies p h h'))
+	    (ensures (r_inv h' v /\ r_repr h v == r_repr h' v))) ->
+
+    /// Initialization (allocation)
+    // A representation value for the initial value of type `a`
+    irepr: Ghost.erased repr ->
+
+    // A property that should hold for all initial values of type `a`.
+    r_init_p: (a -> GTot Type0) ->
+
+    // An initialization operation. We might have a multiple ways of
+    // initializing a given target type `a`; then multiple typeclass instances
+    // should be defined, and each of them can be used properly.
+    r_init: (r:erid ->
```

Comment:
Hello Joonwon, Tahina and I are starting a quick review of this... sorry for the delay.
```suggestion
    r_alloc: (r:erid ->
```

Reply:
The rationale being that we generally use `init` for things that have been allocated (e.g. un-initialized buffers) already but for which we haven't written the initial state.